### PR TITLE
download file when submission is single source

### PR DESCRIPTION
**Pull request by as6325400**
When there is only one uploaded file, download the file directly instead of downloading a compressed file.

### CODE REVIEW

File: webapp/src/Service/SubmissionService.php [hunk 1]
```diff
@@ -768,4 +768,33 @@ public function getSubmissionZipResponse(Submission $submission): StreamedRespon
...
+        }
+
+        $file = $files[0];
+        $filename = $file->getFilename();
+        $sourceCode = $file->getSourcecode();
+
+        return new StreamedResponse(function () use ($sourceCode) {
+            echo $sourceCode;
+        }, 200, [
+            'Content-Type'        => 'application/octet-stream',
```
Comment (meisterT) [inline webapp/src/Service/SubmissionService.php:788]: I wonder whether this should be `text/plain` or something different?

File: webapp/src/Service/SubmissionService.php [hunk 2]
```diff
@@ -768,4 +768,33 @@ public function getSubmissionZipResponse(Submission $submission): StreamedRespon
...
+            'Content-Disposition' => 'attachment; filename="' . $filename . '"',
+            'Content-Length'      => strlen($sourceCode),
+        ]);
+    }
+
+
+    public function getSubmissionFileNums(Submission $submission): int
```
Comment (meisterT) [inline webapp/src/Service/SubmissionService.php:795]: nit: I would prefer the name `getSubmissionFileCount` (and I assume we can mark it as private)

File: webapp/src/Controller/Team/SubmissionController.php [hunk 3]
```diff
@@ -236,6 +236,10 @@ public function downloadAction(int $submitId): Response
                 $submitId));
         }
 
+        if($this->submissionService->getSubmissionFileNums($submission) == 1){
```
Comment (meisterT) [inline webapp/src/Controller/Team/SubmissionController.php:239]: nit: add spaces before/after the parentheses

File: webapp/src/Service/SubmissionService.php (hunk 2 above)
Comment (as6325400) [reply webapp/src/Service/SubmissionService.php:795]: Ok, Should I modify it once more, or will you handle the changes when merging?

File: webapp/src/Service/SubmissionService.php (hunk 1 above)
Comment (as6325400) [reply webapp/src/Service/SubmissionService.php:788]: application/octet-stream is more commonly used when downloading files, but I just tested it, and the effect here is the same as text/plain.

File: webapp/src/Service/SubmissionService.php [hunk 4]
```diff
@@ -768,4 +768,33 @@ public function getSubmissionZipResponse(Submission $submission): StreamedRespon
...
+
+        return new StreamedResponse(function () use ($sourceCode) {
+            echo $sourceCode;
+        }, 200, [
+            'Content-Type'        => 'text/plain',
+            'Content-Disposition' => 'attachment; filename="' . $filename . '"',
+            'Content-Length'      => strlen($sourceCode),
+        ]);
+    }
+
```
Comment (vmcj) [inline webapp/src/Service/SubmissionService.php:793]:
```suggestion
```
small nitpick

File: webapp/src/Service/SubmissionService.php [hunk 5]
```diff
@@ -768,4 +768,33 @@ public function getSubmissionZipResponse(Submission $submission): StreamedRespon
...
+
+    public function getSubmissionFileCount(Submission $submission): int
+    {
+        $files = $submission->getFiles();
+        return count($files);
```
Comment (vmcj) [inline webapp/src/Service/SubmissionService.php:798]:
```suggestion
        return count($submission->getFiles());
```
I think this is still readable and saves a bit of space on our screen.

File: webapp/src/Controller/Team/SubmissionController.php [hunk 6]
```diff
@@ -236,6 +236,10 @@ public function downloadAction(int $submitId): Response
                 $submitId));
         }
 
+        if ($this->submissionService->getSubmissionFileCount($submission) == 1) {
```
Comment (vmcj) [inline webapp/src/Controller/Team/SubmissionController.php:239]:
```suggestion
        if ($this->submissionService->getSubmissionFileCount($submission) === 1) {
```

If we can be very exact I prefer that. It helps steering PHPStan to better results.

**Review by vmcj — COMMENTED** [review]
This will remove the `x` bit from such files, whether that is fine I leave up to the rest.

**Review by meisterT — APPROVED** [review]

**Pull request merged**